Adds SI and JI header updates

zenmonkeykstop opened the pull request:
## Status

Ready for review 

## Description of Changes

Fixes #6186.
Multiple updates to Apache2 security headers in the SI and JI:

- Adds Cross-Origin-\ * headers - see #6176
- Adds `frame-ancestors` attribute in CSP header - see #6178
- Removes X-Download-Options header - see #6180
- Removes X-XSS-Protections header - see #6182 

## Testing

### fresh install (staging):
- check out this branch
- run `make build-debs && make staging`
- on the app server, in both `/etc/apache2/sites-enabled/source.conf` and `/etc/apache2/sites-enabled/journalist.conf`:
  - [ ] Verify that config lines for both X-Download-Options and X-XSS-Protection have been removed
  - [ ] Verify that `frame-ancestors:  'none'` has been added to the Content-Security-Policy header
  - [ ] Verify that Cross-Origin-Opener-Policy is set to "same-origin"
  - [ ] Verify that Cross-Origin-Embedder-Policy  is set to "same-origin"
  - [ ] Verify that Cross-Origin-Resource-Policy is set to "same-site"
  - [ ] Verify that the header changes above are reflected in Apache's response headers for both web interfaces

### upgrade scenario (prod VMs): @cfm
- Using this branch, follow the [upgrade testing procedure](https://docs.securedrop.org/en/stable/development/upgrade_testing.html) to update a productions system with packages built against the branch.
- on the app server, in both `/etc/apache2/sites-enabled/source.conf` and `/etc/apache2/sites-enabled/journalist.conf`:
  - [ ] Verify that config lines for both X-Download-Options and X-XSS-Protection have been removed
  - [ ] Verify that `frame-ancestors:  'none'` has been added to the Content-Security-Policy header
  - [ ] Verify that Cross-Origin-Opener-Policy is set to "same-origin"
  - [ ] Verify that Cross-Origin-Embedder-Policy  is set to "same-origin"
  - [ ] Verify that Cross-Origin-Resource-Policy is set to "same-site"
  - [ ] Verify that the header changes above are reflected in Apache's response headers for both web interfaces
- on the app server, make backup copies of `/etc/apache2/sites-enabled/{source,journalist}.conf` and then run `sudo dpkg-reconfigure securedroop-app-code`
  - [ ] Verify that the backup conf files and the versions in place after `dpkg-reconfigure` are identical.


## Deployment

- New installs will receive these changes via Ansible (with the postint script function that also applies them effectively being a noop).
- Existing installs will be updated via postint.

## Checklist

### If you made changes to the system configuration:

- [x] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

Choose one of the following:

- [ ] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-docs) for these changes, or will do so later
- [x] I would appreciate help with the documentation (these changes themselves don't need docs, but we should consider refreshing the landing page recommendations)
- [ ] These changes do not require documentation

evilaliv3 committed:
Implement Cross-Origin-Opener-Policy Header

Addresses issue: https://github.com/freedomofpress/securedrop/issues/6176

evilaliv3 committed:
Implement CSP Frame Ancestors Policy

Addresses issue: https://github.com/freedomofpress/securedrop/issues/6178

evilaliv3 committed:
Deprecate usage of X-Download-Options header

Implement proposal: https://github.com/freedomofpress/securedrop/issues/6180

evilaliv3 committed:
Deprecate usage of X-XSS-Protection header

Implement proposal: https://github.com/freedomofpress/securedrop/issues/6182

evilaliv3 committed:
Cross-Origin-Embedder-Policy and Cross-Origin-Resource-Policy headers

Implement proposal: https://github.com/freedomofpress/securedrop/issues/6176

zenmonkeykstop committed:
Added postint function to apply Apache2 header changes

zenmonkeykstop committed:
Updated new header directives to use 'unsuccess unset/always set' pattern.

conorsch commented:
@cfm and I will split up testing here. Nice opportunity for me to try out the logic in https://github.com/cfm/terraform-metal-securedrop-production/

cfm reviewed (commented):
Upgrade scenario checks out:

- [x] Using this branch, follow the [upgrade testing procedure](https://docs.securedrop.org/en/stable/development/upgrade_testing.html) to update a production system with packages built against the branch.

```sh-session
vagrant@app-prod:~$ sudo apt upgrade securedrop-app-code
vagrant@app-prod:~$ apt-cache policy securedrop-app-code
securedrop-app-code:
  Installed: 2.2.0~rc1+focal
  Candidate: 2.2.0~rc1+focal
  Version table:
 *** 2.2.0~rc1+focal 500
        500 http://10.0.1.7 focal/main amd64 Packages
        100 /var/lib/dpkg/status
     2.1.0+focal 500
        500 https://apt.freedom.press focal/main amd64 Packages
```

- [x] on the app server, in both `/etc/apache2/sites-enabled/source.conf` and `/etc/apache2/sites-enabled/journalist.conf`:
  - [x] Verify that config lines for both X-Download-Options and X-XSS-Protection have been removed

```sh-session
vagrant@app-prod:~$ grep -E "(X-Download-Options|X-XSS-Protection)" /etc/apache2/sites-enabled/source.conf /etc/apache2/sites-enabled/journalist.conf | wc -l
0
```

  - [x] Verify that `frame-ancestors:  'none'` has been added to the Content-Security-Policy header

```sh-session
vagrant@app-prod:~$ grep "frame-ancestors" /etc/apache2/sites-enabled/source.conf /etc/apache2/sites-enabled/journalist.conf
/etc/apache2/sites-enabled/source.conf:Header always set Content-Security-Policy "default-src 'none'; script-src 'self'; style-src 'self'; img-src 'self'; font-src 'self'; frame-ancestors 'none';"
/etc/apache2/sites-enabled/journalist.conf:Header always set Content-Security-Policy "default-src 'none'; script-src 'self'; style-src 'self'; img-src 'self'; font-src 'self'; frame-ancestors 'none';"
```

  - [x] Verify that Cross-Origin-Opener-Policy is set to "same-origin"

```sh-session
vagrant@app-prod:~$ grep "Cross-Origin-Opener-Policy" /etc/apache2/sites-enabled/source.conf /etc/apache2/sites-enabled/journalist.conf
/etc/apache2/sites-enabled/source.conf:Header onsuccess unset Cross-Origin-Opener-Policy
/etc/apache2/sites-enabled/source.conf:Header always set Cross-Origin-Opener-Policy "same-origin"
/etc/apache2/sites-enabled/journalist.conf:Header onsuccess unset Cross-Origin-Opener-Policy
/etc/apache2/sites-enabled/journalist.conf:Header always set Cross-Origin-Opener-Policy "same-origin"
```

  - [x] Verify that Cross-Origin-Embedder-Policy  is set to "same-origin"

```sh-session
vagrant@app-prod:~$ grep "Cross-Origin-Embedder-Policy" /etc/apache2/sites-enabled/source.conf /etc/apache2/sites-enabled/journalist.conf
/etc/apache2/sites-enabled/source.conf:Header onsuccess unset Cross-Origin-Embedder-Policy
/etc/apache2/sites-enabled/source.conf:Header always set Cross-Origin-Embedder-Policy "same-origin"
/etc/apache2/sites-enabled/journalist.conf:Header onsuccess unset Cross-Origin-Embedder-Policy
/etc/apache2/sites-enabled/journalist.conf:Header always set Cross-Origin-Embedder-Policy "same-origin"

```

  - [x] Verify that Cross-Origin-Resource-Policy is set to "same-site"

```sh-session
vagrant@app-prod:~$ grep "Cross-Origin-Resource-Policy" /etc/apache2/sites-enabled/source.conf /etc/apache2/sites-enabled/journalist.conf
/etc/apache2/sites-enabled/source.conf:Header onsuccess unset Cross-Origin-Resource-Policy
/etc/apache2/sites-enabled/source.conf:Header always set Cross-Origin-Resource-Policy "same-site"
/etc/apache2/sites-enabled/journalist.conf:Header onsuccess unset Cross-Origin-Resource-Policy
/etc/apache2/sites-enabled/journalist.conf:Header always set Cross-Origin-Resource-Policy "same-site"
```

  - [x] Verify that the header changes above are reflected in Apache's response headers for both web interfaces

Source Interface:

```sh-session
amnesia@amnesia:~/Persistent/securedrop$ curl -I $SOURCE | grep -E "(X-Download-Options|X-XSS-Protection)" | wc -l
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0  4398    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
0
amnesia@amnesia:~/Persistent/securedrop$ curl -I $SOURCE | grep -E "(Content-Security-Policy|Cross-Origin)"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0  4398    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
Cross-Origin-Opener-Policy: same-origin
Cross-Origin-Embedder-Policy: same-origin
Cross-Origin-Resource-Policy: same-site
Content-Security-Policy: default-src 'none'; script-src 'self'; style-src 'self'; img-src 'self'; font-src 'self'; frame-ancestors 'none';
```

Journalist Interface:

```sh-session
amnesia@amnesia:~/Persistent/securedrop$ curl -I $JOURNALIST | grep -E "(X-Download-Options|X-XSS-Protection)" | wc -l
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0  2196    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0
0
amnesia@amnesia:~/Persistent/securedrop$ curl -I $JOURNALIST | grep -E "(Content-Security-Policy|Cross-Origin)"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0  2196    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
Cross-Origin-Opener-Policy: same-origin
Cross-Origin-Embedder-Policy: same-origin
Cross-Origin-Resource-Policy: same-site
Content-Security-Policy: default-src 'none'; script-src 'self'; style-src 'self'; img-src 'self'; font-src 'self'; frame-ancestors 'none';
```

- on the app server, make backup copies of `/etc/apache2/sites-enabled/{source,journalist}.conf` and then run `sudo dpkg-reconfigure securedrop-app-code`
  - [x] Verify that the backup conf files and the versions in place after `dpkg-reconfigure` are identical.
  
```sh-session
vagrant@app-prod:~$ sudo cp /etc/apache2/sites-enabled/source.conf /etc/apache2/sites-enabled/source.conf.bak
vagrant@app-prod:~$ sudo cp /etc/apache2/sites-enabled/journalist.conf /etc/apache2/sites-enabled/journalist.conf.bak
vagrant@app-prod:~$ sudo dpkg-reconfigure securedrop-app-code
vagrant@app-prod:~$ diff /etc/apache2/sites-enabled/source.conf*
vagrant@app-prod:~$ diff /etc/apache2/sites-enabled/journalist.conf*
vagrant@app-prod:~$ 
```

Commit:
Postinst apache headers: fail gracefully

The postinst logic for the app-code deb package assumes that the Apache
vhost files exist. That's true on prod, but unfortunately not in
staging, where we sideload locally-build deb pkgs prior to running
the "app" config role at install time.

conorsch commented:
Testing with staging scenario failed: the postinst logic tries to edit the Apache vhost files on disk, but during a staging fresh install, those files don't exist yet. They do in prod, but still, we should fix before merging. I'm testing a patch locally that simply checks for existence first. Based on that, I'll confirm headers on the Onion URLs to finalize review.

conorsch commented:
After appending this fix:

```diff
--- a/install_files/securedrop-app-code/debian/postinst
+++ b/install_files/securedrop-app-code/debian/postinst
@@ -119,6 +119,12 @@ update_to_tls13(){
 #
 # Q4-2021 Apache2 header updates:
 update_apache2_headers(){
+
+    if [ ! -e "$1" ]; then
+        echo "Apache vhost '$1' does not exist, skipping header changes..." >&2
+        return
+    fi
+
     # update CSP header, adding frame-ancestors (see #6178)
     if grep -qP "^Header always set Content-Security-Policy \"default-src 'none'; script-src 'self'; style-src 'self'; img-src 'self'; font-src 'self';\"" "$1"; then
         sed -i "/^Header always set Content-Security-Policy \"default-src 'none'; script-src 'self'; style-src 'self'; img-src 'self'; font-src 'self';\"/c\Header always set Content-Security-Policy \"default-src 'none'; script-src 'self'; style-src 'self'; img-src 'self'; font-src 'self'; frame-ancestors 'none';\"" "$1"
```

the package installs on staging, and the headers are as expected when accessing the Source Interface in Tor Browser. However, I seem to have broken CI with that addition: https://app.circleci.com/pipelines/github/freedomofpress/securedrop/3327/workflows/10261979-1b98-40b9-a3a4-4d5a003f0e8b/jobs/58273 and I'm not yet sure why. Will proceed with testing the Journalist Interface, then poke at CI.

conorsch commented:
The CI error was a failure to rebase, due to a legitimate conflict caused by merge of #6185. I'll rebase manually to get CI passing.

Commit:
Fix tests for CSP Frame Ancestors Policy

The testinfra vars for CSP settings was missing the frame-ancestors
update, due to a wonky rebase by yours truly during review. Tacking on
an extra commit to address, and make the diff match what was in the
already-reviewed PR.

conorsch commented:
That was a rather gnarly rebase, but I think I've got it sorted. Looks to my eye like we have a mix of merges (presumably with conflict resolution) and then I naively tried to rebase that changeset on top of develop, and it unsurprisingly failed. Here's a screenshot of the commit history on this PR today, in the state I reviewed it in:

![gh-merge-rebase-apache-headers-pr](https://user-images.githubusercontent.com/657862/145489256-f9582f00-fc2a-4b7d-96a6-f68860540aa1.png)

I've got a separate branch now that I believe is identical to the patch presented here, and which I propose we force-push prior to approval and merge. However, since I made a lot of destructive changes locally in order to create that PR, I'd appreciate a second set of eyes confirming that the diffs are the same. Pull up these views side by side and verify they're identical:

1. https://github.com/freedomofpress/securedrop/compare/header-updates?expand=1
2. https://github.com/freedomofpress/securedrop/compare/header-updates-rebased-theirs?expand=1

If I can get a +1 on that, then I'm comfortable approving, based on the functional testing I performed today. 

P.S. I made sure to preserve the original branch in https://github.com/freedomofpress/securedrop/compare/header-updates-orig?expand=1 as a precaution.

cfm commented:
@conorsch, I've taken a slightly different approach than you outlined, opting to treat this as a diff of a diff:

```sh-session
cfm@ozymandias securedrop % git checkout header-updates
Branch 'header-updates' set up to track remote branch 'header-updates' from 'origin'.
Switched to a new branch 'header-updates'
cfm@ozymandias securedrop % git diff develop > "$(git branch --show-current).patch"
cfm@ozymandias securedrop % git checkout header-updates-rebased-theirs 
Branch 'header-updates-rebased-theirs' set up to track remote branch 'header-updates-rebased-theirs' from 'origin'.
Switched to a new branch 'header-updates-rebased-theirs'
cfm@ozymandias securedrop % git diff develop > "$(git branch --show-current).patch"
cfm@ozymandias securedrop % diffoscope header-updates.patch header-updates-rebased-theirs.patch --html header-updates.html --text header-updates.patch
```

The resulting [diff](https://gist.github.com/cfm/65e57513b526db16741dbffd2e2aee7e) matches what's expected from rebasing from `develop` after #6190.

As discussed, I'll next force-push from `header-updates-rebased-theirs`, which should retain the passing CI status on aaa4a667aaac63e214642e5008d28f4a13942984, and then approve for merge.

cfm reviewed (approved):
With CI green, approving:

1. functionally based on my <https://github.com/freedomofpress/securedrop/pull/6187#pullrequestreview-825929943> and @conorsch's <https://github.com/freedomofpress/securedrop/pull/6187#issuecomment-990381572>;
2. diffily based on my <https://github.com/freedomofpress/securedrop/pull/6187#issuecomment-990482603>; and, just for good measure,
3. visually based on no surprises flagged by GitHub as **Changed since last view** on the **Files changed** tab.
